plugins/algoltek-usbcr/fu-algoltek-usbcr-device.c

```diff
@@ -0,0 +1,671 @@
+		g_prefix_error(error, "failed to read version: ");
+		return FALSE;
+	}
+	if (!fu_memread_uint16_safe(ver_array->data,
+				    ver_array->len,
+				    130,
+				    &app_ver,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	fu_device_set_version_raw(FU_DEVICE(self), app_ver);
+	if (!fu_memread_uint16_safe(ver_array->data,
+				    ver_array->len,
+				    132,
+				    &boot_ver,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	fu_device_set_version_bootloader_raw(FU_DEVICE(self), boot_ver);
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_probe(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbcrDevice *self = FU_ALGOLTEK_USBCR_DEVICE(device);
+
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usbcr_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	if (g_strcmp0(fu_udev_device_get_devtype(FU_UDEV_DEVICE(self)), "disk") != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "is not correct devtype=%s, expected disk",
+			    fu_udev_device_get_devtype(FU_UDEV_DEVICE(self)));
+		return FALSE;
+	}
+
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "usb", error);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbcrDevice *self = FU_ALGOLTEK_USBCR_DEVICE(device);
+
+	if (!fu_algoltek_usbcr_device_ensure_version(self, error))
+		return FALSE;
+	fu_device_build_vendor_id_u16(device,
+				      "DISK",
+				      fu_udev_device_get_vendor(FU_UDEV_DEVICE(self)));
+	fu_device_set_vendor(FU_DEVICE(self), "Algoltek");
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_algoltek_usbcr_device_prepare_firmware(FuDevice *device,
+					  GInputStream *stream,
+					  FuProgress *progress,
+					  FwupdInstallFlags flags,
+					  GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_algoltek_usbcr_firmware_new();
+
+	/* validate compatibility */
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+	if (fu_algoltek_usbcr_firmware_get_boot_ver(FU_ALGOLTEK_USBCR_FIRMWARE(firmware)) !=
+	    fu_device_get_version_bootloader_raw(device)) {
+		g_set_error(
+		    error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_INVALID_FILE,
+		    "firmware boot version is 0x%x while expecting value is 0x%x",
+		    fu_algoltek_usbcr_firmware_get_boot_ver(FU_ALGOLTEK_USBCR_FIRMWARE(firmware)),
+		    (guint)fu_device_get_version_bootloader_raw(device));
+		return NULL;
+	}
+
+	/* success */
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_write_chunk(FuAlgoltekUsbcrDevice *self, FuChunk *chk, GError **error)
+{
+	guint8 back_data[8] = {0};
+
+	if (!fu_algoltek_usbcr_device_do_write_spi(self,
+						   fu_chunk_get_address(chk),
+						   fu_chunk_get_data(chk),
+						   fu_chunk_get_data_sz(chk),
+						   fu_chunk_get_data_sz(chk),
+						   error))
+		return FALSE;
+	do {
```

Comment:
@lingc24 i think this should use `fu_device_retry_full()` as with this loop it could go on forever if the hardware gets into a bad state.

Reply:
Updated, use fu_device_retry_full() instead of do-while.